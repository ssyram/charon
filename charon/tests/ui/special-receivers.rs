use std::rc::Rc;

trait Simple {
    fn value(&self) -> i32;
}

impl Simple for i32 {
    fn value(&self) -> i32 {
        *self
    }
}

fn test_simple() {
    let rc_val = Rc::new(42i32);
    let trait_obj: Rc<dyn Simple> = rc_val;
    // Actually call the method to trigger vtable code
    let result = trait_obj.value();
    assert_eq!(result, 42);
}

fn main() {
    test_simple();
}