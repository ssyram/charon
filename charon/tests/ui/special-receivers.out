# Final LLBC before serialization:

// Full name: core::fmt::Arguments
#[lang_item("format_arguments")]
pub opaque type Arguments<'a>
where
    'a : 'a,

// Full name: core::marker::MetaSized
#[lang_item("meta_sized")]
pub trait MetaSized<Self>

// Full name: core::marker::Sized
#[lang_item("sized")]
pub trait Sized<Self>
{
    parent_clause0 : [@TraitClause0]: MetaSized<Self>
    non-dyn-compatible
}

// Full name: core::ops::deref::Deref
#[lang_item("deref")]
pub trait Deref<Self>
{
    parent_clause0 : [@TraitClause0]: MetaSized<Self::Target>
    type Target
    fn deref<'_0> = core::ops::deref::Deref::deref<'_0_0, Self>[Self]
    vtable: core::ops::deref::Deref::{vtable}<Self::Target>
}

#[lang_item("deref_method")]
pub fn core::ops::deref::Deref::deref<'_0, Self>(@1: &'_0 (Self)) -> &'_0 (@TraitClause0::Target)
where
    [@TraitClause0]: Deref<Self>,

// Full name: core::ops::drop::Drop
#[lang_item("drop")]
pub trait Drop<Self>
{
    parent_clause0 : [@TraitClause0]: MetaSized<Self>
    fn drop<'_0> = core::ops::drop::Drop::drop<'_0_0, Self>[Self]
    vtable: core::ops::drop::Drop::{vtable}
}

pub fn core::ops::drop::Drop::drop<'_0, Self>(@1: &'_0 mut (Self))
where
    [@TraitClause0]: Drop<Self>,

// Full name: core::option::Option
#[lang_item("Option")]
pub enum Option<T>
where
    [@TraitClause0]: Sized<T>,
{
  None,
  Some(T),
}

// Full name: core::panicking::AssertKind
pub enum AssertKind {
  Eq,
  Ne,
  Match,
}

// Full name: alloc::alloc::Global
#[lang_item("global_alloc_ty")]
pub struct Global {}

// Full name: alloc::rc::Rc
#[lang_item("Rc")]
pub opaque type Rc<T>
where
    [@TraitClause0]: MetaSized<T>,
    [@TraitClause1]: Sized<type_error("removed allocator parameter")>,

// Full name: alloc::rc::{Rc<T>[@TraitClause0::parent_clause0, Sized<Global>]}::new
pub fn new<T>(@1: T) -> Rc<T>[@TraitClause0::parent_clause0, Sized<Global>]
where
    [@TraitClause0]: Sized<T>,

// Full name: alloc::rc::{impl Deref for Rc<T>[@TraitClause0, @TraitClause1]}::deref
pub fn {impl Deref for Rc<T>[@TraitClause0, @TraitClause1]}::deref<'_0, T, A>(@1: &'_0 (Rc<T>[@TraitClause0, @TraitClause1])) -> &'_0 (T)
where
    [@TraitClause0]: MetaSized<T>,
    [@TraitClause1]: Sized<A>,

// Full name: alloc::rc::{impl Deref for Rc<T>[@TraitClause0, @TraitClause1]}
impl<T, A> Deref for Rc<T>[@TraitClause0, @TraitClause1]
where
    [@TraitClause0]: MetaSized<T>,
    [@TraitClause1]: Sized<A>,
{
    parent_clause0 = @TraitClause0
    type Target = T
    fn deref<'_0> = {impl Deref for Rc<T>[@TraitClause0, @TraitClause1]}::deref<'_0_0, T, A>[@TraitClause0, @TraitClause1]
    vtable: {impl Deref for Rc<T>[@TraitClause0, @TraitClause1]}::{vtable}<T, A>[@TraitClause0, @TraitClause1]
}

// Full name: alloc::rc::{impl Drop for Rc<T>[@TraitClause0, @TraitClause1]}::drop
pub fn {impl Drop for Rc<T>[@TraitClause0, @TraitClause1]}::drop<'_0, T, A>(@1: &'_0 mut (Rc<T>[@TraitClause0, @TraitClause1]))
where
    [@TraitClause0]: MetaSized<T>,
    [@TraitClause1]: Sized<A>,

// Full name: alloc::rc::{impl Drop for Rc<T>[@TraitClause0, @TraitClause1]}
impl<T, A> Drop for Rc<T>[@TraitClause0, @TraitClause1]
where
    [@TraitClause0]: MetaSized<T>,
    [@TraitClause1]: Sized<A>,
{
    parent_clause0 = MetaSized<Rc<T>[@TraitClause0, @TraitClause1]>
    fn drop<'_0> = {impl Drop for Rc<T>[@TraitClause0, @TraitClause1]}::drop<'_0_0, T, A>[@TraitClause0, @TraitClause1]
    vtable: {impl Drop for Rc<T>[@TraitClause0, @TraitClause1]}::{vtable}<T, A>[@TraitClause0, @TraitClause1]
}

struct test_crate::Simple::{vtable} {
  size: usize,
  align: usize,
  drop: fn(*mut (dyn exists<_dyn> [@TraitClause0]: Simple<_dyn> + _dyn : '_)),
  method_value: fn<'_0>(&'_0_0 ((dyn exists<_dyn> [@TraitClause0]: Simple<_dyn> + _dyn : '_))) -> i32,
  super_trait_0: &'static (core::marker::MetaSized::{vtable}),
}

// Full name: test_crate::Simple
trait Simple<Self>
{
    parent_clause0 : [@TraitClause0]: MetaSized<Self>
    fn value<'_0> = test_crate::Simple::value<'_0_0, Self>[Self]
    vtable: test_crate::Simple::{vtable}
}

fn test_crate::Simple::value<'_0, Self>(@1: &'_0 (Self)) -> i32
where
    [@TraitClause0]: Simple<Self>,

// Full name: test_crate::{impl Simple for i32}::value
fn {impl Simple for i32}::value<'_0>(@1: &'_0 (i32)) -> i32
{
    let @0: i32; // return
    let self@1: &'_ (i32); // arg #1

    @0 := copy (*(self@1))
    return
}

// Full name: test_crate::{impl Simple for i32}
impl Simple for i32 {
    parent_clause0 = MetaSized<i32>
    fn value<'_0> = {impl Simple for i32}::value<'_0_0>
    vtable: {impl Simple for i32}::{vtable}
}

// Full name: test_crate::test_simple
fn test_simple()
{
    let @0: (); // return
    let rc_val@1: Rc<i32>[MetaSized<i32>, Sized<Global>]; // local
    let trait_obj@2: Rc<(dyn exists<_dyn> [@TraitClause0]: Simple<_dyn> + _dyn : '_)>[MetaSized<(dyn exists<_dyn> [@TraitClause0]: Simple<_dyn> + _dyn : '_)>, Sized<Global>]; // local
    let @3: Rc<i32>[MetaSized<i32>, Sized<Global>]; // anonymous local
    let result@4: i32; // local
    let @5: &'_ ((dyn exists<_dyn> [@TraitClause0]: Simple<_dyn> + _dyn : '_)); // anonymous local
    let @6: &'_ ((dyn exists<_dyn> [@TraitClause0]: Simple<_dyn> + _dyn : '_)); // anonymous local
    let @7: &'_ (Rc<(dyn exists<_dyn> [@TraitClause0]: Simple<_dyn> + _dyn : '_)>[MetaSized<(dyn exists<_dyn> [@TraitClause0]: Simple<_dyn> + _dyn : '_)>, Sized<Global>]); // anonymous local
    let @8: (); // anonymous local
    let @9: (&'_ (i32), &'_ (i32)); // anonymous local
    let @10: &'_ (i32); // anonymous local
    let @11: &'_ (i32); // anonymous local
    let @12: i32; // anonymous local
    let left_val@13: &'_ (i32); // local
    let right_val@14: &'_ (i32); // local
    let @15: bool; // anonymous local
    let @16: i32; // anonymous local
    let @17: i32; // anonymous local
    let kind@18: AssertKind; // local
    let @19: AssertKind; // anonymous local
    let @20: &'_ (i32); // anonymous local
    let @21: &'_ (i32); // anonymous local
    let @22: &'_ (i32); // anonymous local
    let @23: &'_ (i32); // anonymous local
    let @24: Option<Arguments<'_>>[Sized<Arguments<'_>>]; // anonymous local

    storage_live(kind@18)
    storage_live(@19)
    storage_live(@20)
    storage_live(@21)
    storage_live(@22)
    storage_live(@23)
    storage_live(@24)
    storage_live(rc_val@1)
    rc_val@1 := new<i32>[Sized<i32>](const (42 : i32))
    storage_live(trait_obj@2)
    storage_live(@3)
    @3 := move (rc_val@1)
    trait_obj@2 := unsize_cast<Rc<i32>[MetaSized<i32>, Sized<Global>], Rc<(dyn exists<_dyn> [@TraitClause0]: Simple<_dyn> + _dyn : '_)>[MetaSized<(dyn exists<_dyn> [@TraitClause0]: Simple<_dyn> + _dyn : '_)>, Sized<Global>]>(move (@3))
    drop[{impl Drop for Rc<T>[@TraitClause0, @TraitClause1]}<i32, Global>[MetaSized<i32>, Sized<Global>]] @3
    storage_dead(@3)
    storage_live(result@4)
    storage_live(@5)
    storage_live(@6)
    storage_live(@7)
    // Actually call the method to trigger vtable code
    @7 := &trait_obj@2
    @6 := {impl Deref for Rc<T>[@TraitClause0, @TraitClause1]}::deref<'_, (dyn exists<_dyn> [@TraitClause0]: Simple<_dyn> + _dyn : '_), Global>[MetaSized<(dyn exists<_dyn> [@TraitClause0]: Simple<_dyn> + _dyn : '_)>, Sized<Global>](move (@7))
    @5 := &*(@6)
    storage_dead(@7)
    result@4 := Simple<(dyn exists<_dyn> [@TraitClause0]: Simple<_dyn> + _dyn : '_)>::value<'_>(move (@5))
    storage_dead(@5)
    storage_dead(@6)
    storage_live(@8)
    storage_live(@9)
    storage_live(@10)
    @10 := &result@4
    storage_live(@11)
    storage_live(@12)
    @12 := const (42 : i32)
    @11 := &@12
    @9 := (move (@10), move (@11))
    storage_dead(@11)
    storage_dead(@10)
    storage_live(left_val@13)
    left_val@13 := copy ((@9).0)
    storage_live(right_val@14)
    right_val@14 := copy ((@9).1)
    storage_live(@15)
    storage_live(@16)
    @16 := copy (*(left_val@13))
    storage_live(@17)
    @17 := copy (*(right_val@14))
    @15 := move (@16) == move (@17)
    if move (@15) {
    }
    else {
        storage_dead(@17)
        storage_dead(@16)
        storage_live(kind@18)
        kind@18 := AssertKind::Eq {  }
        storage_live(@19)
        @19 := move (kind@18)
        storage_live(@20)
        storage_live(@21)
        @21 := &*(left_val@13)
        @20 := &*(@21)
        storage_live(@22)
        storage_live(@23)
        @23 := &*(right_val@14)
        @22 := &*(@23)
        storage_live(@24)
        @24 := Option::None {  }
        panic(core::panicking::assert_failed)
    }
    storage_dead(@17)
    storage_dead(@16)
    storage_dead(@15)
    storage_dead(right_val@14)
    storage_dead(left_val@13)
    storage_dead(@12)
    storage_dead(@9)
    storage_dead(@8)
    @0 := ()
    storage_dead(result@4)
    drop[{impl Drop for Rc<T>[@TraitClause0, @TraitClause1]}<(dyn exists<_dyn> [@TraitClause0]: Simple<_dyn> + _dyn : '_), Global>[MetaSized<(dyn exists<_dyn> [@TraitClause0]: Simple<_dyn> + _dyn : '_)>, Sized<Global>]] trait_obj@2
    storage_dead(trait_obj@2)
    drop[{impl Drop for Rc<T>[@TraitClause0, @TraitClause1]}<i32, Global>[MetaSized<i32>, Sized<Global>]] rc_val@1
    storage_dead(rc_val@1)
    @0 := ()
    return
}

// Full name: test_crate::main
fn main()
{
    let @0: (); // return
    let @1: (); // anonymous local

    storage_live(@1)
    @1 := test_simple()
    storage_dead(@1)
    @0 := ()
    @0 := ()
    return
}



